File: src/storage/lance/lance.client.ts
import * as lancedb from '@lancedb/lancedb';
import path from 'path';

const DEFAULT_DB_PATH = path.resolve(process.cwd(), '.lancedb');

/**
 * Manages a shared LanceDB connection with multiple named tables.
 * Tables are lazily created on first write (LanceDB requires data to infer schema).
 */
export class LanceClient {
  private db: lancedb.Connection | null = null;
  private tables: Map<string, lancedb.Table> = new Map();
  private readonly dbPath: string;

  constructor(dbPath: string = DEFAULT_DB_PATH) {
    this.dbPath = dbPath;
  }

  // ──────────────────────────────────────────
  //  Connection
  // ──────────────────────────────────────────

  async connect(): Promise<lancedb.Connection> {
    if (this.db) return this.db;

    this.db = await lancedb.connect(this.dbPath);
    console.log(`[LanceClient] Connected to ${this.dbPath}`);
    return this.db;
  }

  async close(): Promise<void> {
    this.tables.clear();
    this.db = null;
  }

  // ──────────────────────────────────────────
  //  Table Management
  // ──────────────────────────────────────────

  /**
   * Opens an existing table or returns null if it doesn't exist yet.
   * Tables that don't exist will be created lazily when data is first written.
   */
  async openTable(tableName: string): Promise<lancedb.Table | null> {
    if (this.tables.has(tableName)) {
      return this.tables.get(tableName)!;
    }

    const db = await this.connect();
    const existingTables = await db.tableNames();

    if (existingTables.includes(tableName)) {
      const table = await db.openTable(tableName);
      this.tables.set(tableName, table);
      return table;
    }

    return null;
  }

  /**
   * Creates a table with initial data. If the table already exists, opens it instead.
   */
  async createTable(tableName: string, data: Record<string, any>[]): Promise<lancedb.Table> {
    const db = await this.connect();
    const existingTables = await db.tableNames();

    let table: lancedb.Table;
    if (existingTables.includes(tableName)) {
      table = await db.openTable(tableName);
    } else {
      table = await db.createTable(tableName, data);
    }

    this.tables.set(tableName, table);
    return table;
  }

  /**
   * Adds data to an existing table. Creates the table if it doesn't exist.
   */
  async addToTable(tableName: string, data: Record<string, any>[]): Promise<lancedb.Table> {
    let table = await this.openTable(tableName);

    if (!table) {
      // Table doesn't exist — create it with this data
      table = await this.createTable(tableName, data);
    } else {
      await table.add(data);
    }

    return table;
  }

  /**
   * Query rows with a SQL-like filter.
   */
  async filter(tableName: string, where: string, select?: string[]): Promise<any[]> {
    const table = await this.openTable(tableName);
    if (!table) return [];

    try {
      let query = table.query().where(where);
      if (select) query = query.select(select);
      return await query.toArray();
    } catch (err) {
      console.warn(`[LanceClient] Filter error on "${tableName}":`, err);
      return [];
    }
  }

  /**
   * Vector similarity search on a table.
   */
  async vectorSearch(
    tableName: string,
    queryVector: number[],
    topK: number,
    distanceType: 'l2' | 'cosine' | 'dot' = 'cosine',
  ): Promise<any[]> {
    const table = await this.openTable(tableName);
    if (!table) return [];

    return table
      .vectorSearch(queryVector)
      .distanceType(distanceType)
      .limit(topK)
      .toArray();
  }
}

File: pipeline/storage/lance_client.py
"""
LanceDB client for the Python pipeline.
Mirrors the TS storage/lance/lance.client.ts.
"""

import lancedb
import logging
from pathlib import Path
from typing import Dict, Any, List

logger = logging.getLogger(__name__)

DEFAULT_DB_PATH = Path(__file__).resolve().parent.parent.parent / ".lancedb"

class LanceClient:
    """Wrapper around lancedb for Python."""
    
    def __init__(self, db_path: Path | str = DEFAULT_DB_PATH):
        self.db_path = str(db_path)
        self.db = None
        
    def connect(self):
        if self.db is None:
            self.db = lancedb.connect(self.db_path)
            logger.info(f"[LanceClient] Connected to {self.db_path}")
        return self.db

    def close(self):
        self.db = None

    def open_table(self, table_name: str):
        db = self.connect()
        if table_name in db.table_names():
            return db.open_table(table_name)
        return None

    def create_table(self, table_name: str, data: List[Dict[str, Any]]):
        db = self.connect()
        if table_name in db.table_names():
            return db.open_table(table_name)
        return db.create_table(table_name, data=data)

    def add_to_table(self, table_name: str, data: List[Dict[str, Any]]):
        if not data:
            return None
            
        table = self.open_table(table_name)
        if table is None:
            return self.create_table(table_name, data)
        else:
            table.add(data)
            return table

    def vector_search(self, table_name: str, query_vector: List[float], top_k: int = 10, distance_type: str = "cosine"):
        table = self.open_table(table_name)
        if table is None:
            return []
            
        return table.search(query_vector).metric(distance_type).limit(top_k).to_list()
